src/porepy/models/abstract_model.py

```diff
@@ -1,37 +1,87 @@
-""" Contains the mothed class for all models, that is, standardized setups for complex
+""" Contains the mother class for all models, that is, standardized setups for complex
 problems.
 
+Some methods may raise NotImplementedErrors instead of being abstract methods. This
+is because they are only needed for some model types (typically nonlinear ones).
+
+Note on running simulations: We provide run_time_dependent_model and run_stationary_model.
+Both call the Model's prepare_simulation and after_simulation methods, as well as the
+solve method of either a LinearSolver or NewtonSolver (only non-linear available
+for the time being). The Solvers call methods before_newton_loop, check_convergence,
+after_newton_convergence/after_newton_divergence, (only NewtonSolver:) before_newton_iteration,
+after_newton_iteration. The name "newton" is there for legacy reasons, a more fitting
+and general name would be something like "equation_solve".
 """
 import abc
-from typing import Any, Dict, Tuple
+import logging
+from typing import Any, Dict, Optional, Tuple
 
 import numpy as np
 
 import porepy as pp
 
+logger = logging.getLogger(__name__)
 module_sections = ["models", "numerics"]
 
 
-class AbstractModel(abc.ABC):
-    """This is an abstract class that specifies methods that a model must implement to
+class AbstractModel:
+    """This is a class that specifies methods that a model must implement to
     be compatible with the Newton and time stepping methods.
 
     """
 
-    @abc.abstractmethod
-    @pp.time_logger(sections=module_sections)
-    def get_state_vector(self) -> np.ndarray:
-        """Get a vector of the current state of the variables; with the same ordering
-            as in the assembler.
-
-        Returns:
-            np.array: The current state of the system.
-
+    def __init__(self, params: Optional[Dict] = None):
+        if params is None:
+            self.params = {}
+        else:
+            self.params = params
+        if params is None:
+            params = {}
+        default_params = {
+            "folder_name": "visualization",
+            "file_name": "data",
+            "use_ad": False,
+        }
+        default_params.update(params)
+        self.params = default_params
+
+        # Set a convergence status. Not sure if a boolean is sufficient, or whether
+        # we should have an enum here.
+        self.convergence_status: bool = False
+        self.linear_solver: str = "direct"
+
+        self._iteration: int = 0
```

Comment:
Non-linear. It seems to me that we don't use it (except updating the counter in after_newton_iteration). However, it may be really useful. Should we rename it to _nonlinear_iteration and keep it?

Reply:
Yes please.